### ISSUE TITLE
model loading gets into swap

### ISSUE THREAD

**Issue by kuanjames** [jundot/omlx#429]
if memory guard is on, when the server starts or when i load a model, it uses way too much vram, i.e. more than it needs. swap is being used. swap is only partially released after the model has loaded into the vram.

this problem does not occur when memory guard is off.

the problem is, memory guard turns itself back on everytime i restart the app or simply refresh the webpage.

i am on v.0.2.23

**Comment by wsantos** [comment]
Something weird is happening at first load, when I load `Qwen3.5-35B-A3B-oQ3 (15.62 GB)`, my memory usage spikes to 23GB and goes near 16GB after the first prompt, this might be related

**Comment by wsantos** [comment]
More info @jundot : 

after load and idle:

<img width="1340" height="235" alt="Image" src="https://github.com/user-attachments/assets/6b94c6ef-f9b8-481f-b55f-dc013428b751" />

after the first message:

<img width="1219" height="469" alt="Image" src="https://github.com/user-attachments/assets/e708d55e-e983-412e-93dd-f898b5359121" />

**Comment by jundot** [comment]
@kuanjames @wsantos Both issues are fixed in [v0.2.24.dev2](https://github.com/jundot/omlx/releases/tag/v0.2.24.dev2):

**1. ~2x memory usage after model load**

`mx.set_cache_limit(total_mem)` at startup tells MLX to keep all freed Metal buffers in a pool for reuse. The problem was that intermediate tensors from model loading (weight transforms, quantization) stayed in that pool indefinitely because `mx.clear_cache()` was never called after load. The first inference request would eventually clear them, but until then memory sat at ~2x model size. Fixed by calling `mx.synchronize() + mx.clear_cache()` right after model load completes.

**2. Memory guard toggle resetting on refresh**

The `GET /admin/api/global-settings` endpoint was returning `max_process_memory` but not `prefill_memory_guard`. The frontend initializes with `prefill_memory_guard: true` as default, and since the API never sent the saved value back, every page load reverted to the default. The setting was being saved to disk correctly, just never read back by the frontend. Fixed by adding `prefill_memory_guard` to the API response.

**Comment by kuanjames** [comment]
@jundot thanks for the update.

However, these are still happening:

1. Start app --> model starts loading and runs into SWAP again.
2. Stop server then restart server --> Same SWAP problem

But this was fixed:

3. Start app, get hit with the SWAP problem and wait for it to be released from SWAP, then unload model and load again --> no new SWAP problem.

I hope this helps.

**Comment by kuanjames** [comment]
Just tried to unpin all models, restart app... then manually load a model... SWAP problem happens.

**Comment by jundot** [comment]
@kuanjames I found the root cause of the swap issue. `mx.set_memory_limit()` in the memory guard was causing MLX to aggressively reclaim cached buffers during model loading, creating alloc/free churn that pushes the system into swap. This explains why memory guard OFF had no problem — the metal-level limits were never set.

Removed the `mx.set_memory_limit()` and `mx.set_cache_limit()` calls entirely. All the actual memory protection (model eviction, prefill memory check, generation guard) uses `mx.get_active_memory()` polling and doesn't depend on these Metal-level limits at all.

Could you test with dev3?
- macOS 15 (Sequoia): https://omlx.ai/beta/oMLX-0.2.24.dev3-macos15-sequoia_260328112502.dmg
- macOS 26 (Tahoe): https://omlx.ai/beta/oMLX-0.2.24.dev3-macos26-tahoe_260328112728.dmg

All three issues should be fixed now:
1. Swap during model load → removed metal-level limits
2. ~2x memory after load → clear Metal pool after load completes
3. Memory guard toggle resetting → API now returns saved value

**Comment by kuanjames** [comment]
@jundot I just tried it. Sorry, same SWAP problems as dev2. I tried all those scenarios too. Same effect.

**Comment by kuanjames** [comment]
Here is the log, not sure it helps

026-03-28 10:43:54,067 - omlx.engine_pool - INFO - [-] - Discovered 7 models, max memory: disabled
2026-03-28 10:43:54,067 - omlx.engine_pool - INFO - [-] - Applied model_type override for Qwen3-Embedding-0.6B-4bit-DWQ: type=embedding, engine=embedding
2026-03-28 10:43:54,067 - omlx.server_metrics - INFO - [-] - Loaded all-time stats from /Users/xxxx/.omlx/stats.json
2026-03-28 10:43:54,067 - omlx.server - INFO - [-] - Server initialized with 7 models
2026-03-28 10:43:54,067 - omlx.server - INFO - [-] - Default model: Qwen3.5-122B-A10B-oQ4
2026-03-28 10:43:54,067 - omlx.server - INFO - [-] - Max model memory: disabled (no limit)
2026-03-28 10:43:54,067 - omlx.server - INFO - [-] - Default max tokens: 32768
2026-03-28 10:43:54,067 - omlx.server - INFO - [-] - API key authentication: enabled
2026-03-28 10:43:54,068 - omlx.server - INFO - [-] - HF Downloader initialized
2026-03-28 10:43:54,120 - omlx.server - INFO - [-] - ModelScope Downloader initialized
2026-03-28 10:43:54,121 - omlx.server - INFO - [-] - oQ Quantizer initialized
2026-03-28 10:43:54,122 - omlx.server - INFO - [-] - HF Uploader initialized
2026-03-28 10:43:54,123 - asyncio - DEBUG - [-] - Using selector: KqueueSelector
2026-03-28 10:44:00,092 - omlx.engine_pool - INFO - [-] - Loading model: Qwen3.5-122B-A10B-oQ4
2026-03-28 10:44:00,192 - omlx.engine.vlm - DEBUG - [-] - Removed video_processor from MODALITY_TO_AUTOPROCESSOR_MAPPING
2026-03-28 10:44:12,006 - omlx.engine_pool - WARNING - [-] - VLM loading failed for Qwen3.5-122B-A10B-oQ4, falling back to LLM: 
Qwen2VLImageProcessor requires the Torchvision library but it was not found in your environment. Check out the instructions on the
installation page: https://pytorch.org/get-started/locally/ and follow the ones that match your environment.
Please note that you may need to restart your runtime after installation.

Qwen2VLImageProcessor requires the PyTorch library but it was not found in your environment. Check out the instructions on the
installation page: https://pytorch.org/get-started/locally/ and follow the ones that match your environment.
Please note that you may need to restart your runtime after installation.

2026-03-28 10:44:12,006 - omlx.engine.vlm - INFO - [-] - VLMBatchedEngine stopped
2026-03-28 10:44:12,033 - omlx.utils.tokenizer - DEBUG - [-] - Qwen3 detected: setting eos_token to <|im_end|>
2026-03-28 10:44:45,452 - omlx.model_registry - DEBUG - [-] - Engine eb83be05-2c66-4f6b-9aa4-ac554ed1b91f acquired model 4643471632
2026-03-28 10:44:45,739 - omlx.scheduler - INFO - [-] - Loaded 1 additional EOS token(s) from generation_config.json: {248044}
2026-03-28 10:44:45,740 - omlx.scheduler - INFO - [-] - Enlarging paged cache block_size=256 to 2048 for ArraysCache hybrid model (reduces boundary snapshot overhead)
2026-03-28 10:44:45,740 - omlx.scheduler - INFO - [-] - paged SSD-only mode: max_blocks=100000, block_size=2048 tokens
2026-03-28 10:44:45,741 - omlx.cache.paged_cache - INFO - [-] - PagedCacheManager initialized: block_size=2048, initial_blocks=256, max_blocks=100000, max_tokens=204800000
2026-03-28 10:44:45,742 - omlx.cache.paged_ssd_cache - INFO - [-] - Scanning SSD cache directory: /Volumes/Cache
2026-03-28 10:44:45,824 - omlx.cache.paged_ssd_cache - INFO - [-] - SSD cache scan complete: scanned=145, indexed=145, errors=0, total_size=17.61 GB
2026-03-28 10:44:45,824 - omlx.cache.paged_ssd_cache - INFO - [-] - PagedSSDCacheManager initialized: dir=/Volumes/Cache, max_size=465.00 GB, existing_files=145, disk_free=448.00 GB, cache_used=17.61 GB
2026-03-28 10:44:45,824 - omlx.cache.paged_cache - INFO - [-] - paged SSD cache manager connected to PagedCacheManager
2026-03-28 10:44:45,824 - omlx.cache.prefix_cache - INFO - [-] - PagedSSDCacheManager connected to BlockAwarePrefixCache
2026-03-28 10:44:45,824 - omlx.scheduler - INFO - [-] - paged SSD cache enabled: cache_dir=/Volumes/Cache, max_size=465.00 GB, block_size=2048 tokens
2026-03-28 10:44:45,824 - omlx.scheduler - INFO - [-] - paged SSD cache enabled: /Volumes/Cache, block_size=2048, max_blocks=100000
2026-03-28 10:44:45,824 - omlx.engine_core - DEBUG - [-] - Engine eb83be05-2c66-4f6b-9aa4-ac554ed1b91f initialized
2026-03-28 10:44:45,825 - omlx.engine_core - INFO - [-] - Engine started
2026-03-28 10:44:45,825 - omlx.engine.batched - INFO - [-] - BatchedEngine loaded: /Users/xxxx.lmstudio/models/Qwen3.5-122B-A10B-oQ4
2026-03-28 10:44:45,825 - omlx.engine_pool - INFO - [-] - Successfully loaded Qwen3.5-122B-A10B-oQ4 as LLM (fallback from VLM)
2026-03-28 10:44:46,921 - omlx.engine_pool - INFO - [-] - Loaded model: Qwen3.5-122B-A10B-oQ4 (estimated: 69.87GB, total: 69.87GB)
2026-03-28 10:44:46,921 - omlx.admin.routes - INFO - [-] - Manually loaded model: Qwen3.5-122B-A10B-oQ4

**Comment by jundot** [comment]
@kuanjames Thanks for the log, that's very helpful.

I noticed the VLM fallback happening — `Qwen2VLImageProcessor requires the Torchvision library` causes the VLM load to fail, then it reloads as LLM. This means your 70GB model gets loaded twice: once as VLM (12 seconds, weights already in memory when the processor import fails), then again as LLM (33 seconds). Even though there's a `gc.collect() + mx.clear_cache()` in between, if the GC doesn't fully release the first load's tensors (circular references), peak memory could hit ~140GB before settling back down.

This VLM fallback path exists in v0.2.20 as well, so it doesn't explain the memory guard ON/OFF difference by itself. But it's likely contributing to the swap pressure.

A couple of things that would help narrow this down:

1. What's your hardware spec? (chip, RAM size)
2. When you tested "memory guard OFF = no swap", was that with the same Qwen3.5-122B-A10B-oQ4 model (70GB)? Or a different/smaller model?
3. In the dev3 logs, do you see a line that says `Metal memory limit set:`? (You should NOT see it in dev3 — if you do, it means the old build is still running)
4. Could you try one of these to isolate the VLM double-load from the swap issue?
   - Load a model that doesn't trigger VLM fallback (e.g. a plain mlx-4bit model)
   - Or go to the model's settings in admin and change model type to "LLM" — this skips the VLM attempt entirely

**Comment by jundot** [comment]
@kuanjames Looking at the logs more carefully, the VLM fallback on your Qwen3.5-122B-A10B-oQ4 is abnormal. Normally it should load as VLM successfully like this:

```
Using `use_fast=True` but `torchvision` is not available. Falling back to the slow image processor.
```

But your log shows a hard failure with `Qwen2VLImageProcessor requires the Torchvision library`. I'm tracking this separately in #431.

For now, could you set model type to "LLM" in the model settings page and test again? This skips the VLM attempt entirely and avoids the double-load. It'll help isolate whether the swap is from the VLM fail → LLM retry or something else.

**Comment by kuanjames** [comment]
Yes. Changing it to LLM fixed it.

Why is that? This was not a problem with v0.2.20

Edit: read the other post. I see the issue now.

**Comment by jundot** [comment]
@kuanjames Found the root cause of the VLM fallback that was causing the double-load and swap pressure.

`transformers` 5.4.0 was released on March 27 and changed `Qwen2VLImageProcessor` from a numpy/PIL backend to a torch/torchvision backend. The module now has `import torch` at the top level, so it can't be imported at all without torch installed. Since oMLX doesn't bundle torch in the DMG, every Qwen3.5 model triggers: VLM load attempt (fails after loading weights) → cleanup → LLM reload. That's your 70GB model loaded twice, exactly what caused the swap spike.

This wasn't happening in v0.2.20 because the bundled transformers version was older. The v0.2.23 build just happened to pull in 5.4.0 since the dependency had no upper bound.

Fix is pinning `transformers>=5.0.0,<5.4.0`. With this, Qwen3.5 models will load as VLM on the first try using the slow image processor (no torch needed), so no more double-load. Pushing a new release shortly.

**Comment by jundot** [comment]
[v0.2.24](https://github.com/jundot/omlx/releases/tag/v0.2.24) is out with all three fixes (Metal limits removed, Metal pool cleared after load, transformers pinned). Thanks a a lot!

**Comment by kuanjames** [comment]
You fixed it! Thank you. It works.
